src/Raven.Server/Documents/Queries/Regex/RegexTermEnum.cs

```diff
@@ -0,0 +1,78 @@
+﻿/* 
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System.Text.RegularExpressions;
+using Lucene.Net.Index;
+using Lucene.Net.Store;
+
+namespace Lucene.Net.Search
+{
+    /// <summary>
+    /// Subclass of FilteredTermEnum for enumerating all terms that match the
+    /// specified regular expression term using the specified regular expression
+    /// implementation.
+    /// <para>Term enumerations are always ordered by Term.compareTo().  Each term in
+    /// the enumeration is greater than all that precede it.</para>
+    /// </summary>
+    /// <remarks>http://www.java2s.com/Open-Source/Java-Document/Net/lucene-connector/org/apache/lucene/search/regex/RegexTermEnum.java.htm</remarks>
+    public class RegexTermEnum : FilteredTermEnum
+    {
+        private string _sField;
+        private bool _bEndEnum;
+        private readonly string _pattern;
+
+        public RegexTermEnum(IndexReader reader, Term term, IState state)
+        {
+            _sField = term.Field;
+
+            _pattern = term.Text;
+
+            SetEnum(reader.Terms(new Term(term.Field, string.Empty), state), state);
+        }
+
+        /// <summary>Equality compare on the term </summary>
+        protected override bool TermCompare(Term term)
+        {
+            if (_sField == term.Field)
+            {
+                return Regex.IsMatch(term.Text, _pattern);
```

Comment:
Have a `Regex` instance that we hold on to.
The logic goes like this.
A concurrent dictionary of up to 1024 items (configuration).
If exceeding that, evict random item from cache.
If cache is full, and number of regex creations per minute exceed (configuration), clear the cache, raise alert, and then move to using non compiled regex.

Reply:
Let's set eviction policy on least recently used and handle too many at the same time.